modules/substrate/src/lib.rs

```diff
@@ -371,27 +375,93 @@ impl<T: Config> bp_header_chain::HeaderChain<BridgedHeader<T>> for Module<T> {
+		}
 
 		Ok(())
 	}
+}
 
-	fn import_finality_proof(header: BridgedHeader<T>, finality_proof: Vec<u8>) -> Result<(), ()> {
-		let mut verifier = verifier::Verifier {
-			storage: PalletStorage::<T>::new(),
-		};
+/// Import a finalized header without checking if this is true.
+///
+/// This function assumes that all the given header has already been proven to be valid and
+/// finalized. Using this assumption it will write them to storage with minimal checks. That
+/// means it's of great importance that this function *not* called with any headers whose
+/// finality has not been checked, otherwise you risk bricking your bridge.
+///
+/// One thing this function does do for you is GRANDPA authority set handoffs. However, since it
+/// does not do verification on the incoming header it will assume that the authority set change
+/// signals in the digest are well formed.
+fn import_header_unchecked<S, T>(storage: &mut S, header: BridgedHeader<T>)
+where
+	S: BridgeStorage<Header = BridgedHeader<T>>,
+	T: Config,
+{
+	// Since we want to use the existing storage infrastructure we need to indicate the fork
+	// that we're on. We will assume that since we are using the unchecked import there are no
+	// forks, and can indicate that by using the first imported header's "fork".
+	let dummy_fork_hash = <InitialHash<T>>::get();
+
+	// If we have a pending change in storage let's check if the current header enacts it.
+	let enact_change = if let Some(pending_change) = storage.scheduled_set_change(dummy_fork_hash) {
+		pending_change.height == *header.number()
+	} else {
+		// We don't have a scheduled change in storage at the moment. Let's check if the current
+		// header signals an authority set change.
+		if let Some(change) = verifier::find_scheduled_change(&header) {
+			let next_set = AuthoritySet {
+				authorities: change.next_authorities,
+				set_id: storage.current_authority_set().set_id + 1,
+			};
 
-		let _ = verifier
-			.import_finality_proof(header.hash(), finality_proof.into())
-			.map_err(|_| ())?;
+			let height = *header.number() + change.delay;
+			let scheduled_change = ScheduledChange {
+				authority_set: next_set,
+				height,
+			};
 
-		Ok(())
+			storage.schedule_next_set_change(dummy_fork_hash, scheduled_change);
+
+			// If the delay is 0 this header will enact the change it signaled
+			height == *header.number()
+		} else {
+			false
+		}
+	};
+
+	if enact_change {
+		const ENACT_SET_PROOF: &str = "We only set `enact_change` as `true` if we are sure that there is a scheduled
+				authority set change in storage. Therefore, it must exist.";
+
+		// If we are unable to enact an authority set it means our storage entry for scheduled
+		// changes is missing. Best to crash since this is likely a bug.
+		let _ = storage.enact_authority_set(dummy_fork_hash).expect(ENACT_SET_PROOF);
 	}
+
+	storage.update_best_finalized(header.hash());
+
+	storage.write_header(&ImportedHeader {
+		header,
+		requires_justification: false,
```

Comment:
Just trying to check that we are ok under all conditions - we're assuming that the header does not `requires_justification`, because we have checked justification in the finality verifier pallet. And we rely on the fact that justification of the transition header must be singed by the current authority set. So if justification is provided for the header from the next session, it will be rejected? Am I right?

So if `H` is transition header, then we will reject transaction with headers `H-2, H-1, H, H+1`, because `H+1` is signed by the next set.

Reply:
In the finality verifier pallet we check the justification against the current authority set known to the bridge pallet. So in the scenario you're describing the authority set which signed the given justification and the current authority set would be different, and we would therefore reject the header.